server_tcp.go

```diff
@@ -42,81 +46,170 @@ func (w *TCPResponseWriter) WriteMsg(m *dns.Msg) error {
-	certTxt := packTxtString(certBuf)
-
-	log.Info("Entering DNSCrypt TCP listening loop tcp://%s", l.Addr().String())
 
-	for {
+	for s.isStarted() {
 		conn, err := l.Accept()
-		if err == nil {
-			go func() {
-				_ = s.handleTCPConnection(conn, certTxt)
-				_ = conn.Close()
-			}()
-		}
 
+		// Check the error code and exit loop if necessary
 		if err != nil {
+			if !s.isStarted() {
+				// Stopped gracefully
+				break
+			}
+			var netErr net.Error
+			if errors.As(err, &netErr) && netErr.Temporary() {
+				// Note that timeout errors will be here (i.e. hitting ReadDeadline)
+				continue
+			}
 			if isConnClosed(err) {
 				log.Info("udpListen.ReadFrom() returned because we're reading from a closed connection, exiting loop")
 			} else {
 				log.Info("got error when reading from UDP listen: %s", err)
 			}
 			break
 		}
+
+		// If we got here, the connection is alive
+		s.lock.Lock()
+		// Track the connection to allow unblocking reads on shutdown.
+		s.tcpConns[conn] = struct{}{}
+		s.lock.Unlock()
+
+		tcpWg.Add(1)
+		go func() {
+			// Ignore error here, it is most probably a legit one
+			// if not, it's written to the debug log
+			_ = s.handleTCPConnection(conn, certTxt)
+
+			// Clean up
+			_ = conn.Close()
+			s.lock.Lock()
+			delete(s.tcpConns, conn)
+			s.lock.Unlock()
+			tcpWg.Done()
+		}()
+	}
+
+	return nil
+}
+
+// prepareServeTCP prepares the server and listener to serving DNSCrypt
+func (s *Server) prepareServeTCP(l net.Listener) error {
+	// Check that server is properly configured
+	if !s.validate() {
+		return ErrServerConfig
 	}
 
+	// Protect shutdown-related fields
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.initOnce.Do(s.init)
+
+	// Mark the server as started if needed
+	s.started = true
+
+	// Track an active TCP listener
+	s.tcpListeners[l] = struct{}{}
 	return nil
 }
 
+// cleanUpTCP waits until all TCP messages before cleaning up
+func (s *Server) cleanUpTCP(tcpWg *sync.WaitGroup, l net.Listener) {
+	// Wait until all TCP connections are processed
+	tcpWg.Wait()
+
+	// Not using it anymore so can be removed from the active listeners
+	s.lock.Lock()
+	delete(s.tcpListeners, l)
+	s.lock.Unlock()
+
+	// The work is finished
+	s.wg.Done()
+}
+
+// handleTCPMsg handles a single TCP message. If this method returns error
+// the connection will be closed
+func (s *Server) handleTCPMsg(b []byte, conn net.Conn, certTxt string) error {
+	if len(b) < minDNSPacketSize {
+		// Ignore the packets that are too short
+		return ErrTooShort
+	}
+
+	if bytes.Equal(b[:clientMagicSize], s.ResolverCert.ClientMagic[:]) {
+		// This is an encrypted message, we should decrypt it
+		m, q, err := s.decrypt(b)
+		if err != nil {
+			return fmt.Errorf("failed to decrypt incoming message: %w", err)
+		}
+		rw := &TCPResponseWriter{
+			tcpConn: conn,
+			encrypt: s.encrypt,
+			req:     m,
+			query:   q,
+		}
+		err = s.serveDNS(rw, m)
+		if err != nil {
+			return fmt.Errorf("failed to process a DNS query: %w", err)
+		}
+	} else {
```

Comment:
If I am not mistaken, you could just `return nil above this line and remove the `else`.